### PR TITLE
chore(logstreams): introduce async snapshotting

### LINKED ISSUE #2292: Investigate how to create snapshots asynchronously

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* created new state machine for the snapshotting (need to document this)
 * snapshot is done independent from the normal stream processing
 * it is ensured that at max 3 valid snapshots are created on runtime - follow up issue should make this configurable
 * need follow up issue to make snapshot time configurable again
 * reprocessing and recovery works the same way

@saig0  lets discuss this tomorrow :) 

closes #2292

### CODE REVIEW

**Comment by Zelldon** [comment]
Changed the PR as discussed, we will now store the latest successful processed event position in RocksDB. This makes it possible to ensure on which position the snapshot was taken and start reprocessing from this point. @saig0 please have a look

File: logstreams/src/test/java/io/zeebe/logstreams/processor/AsyncSnapshotingTest.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            () -> logStream.getCommitPosition());
+    actorScheduler.submitActor(asyncSnapshotDirector).join();
+  }
+
+  @Test
+  public void shouldStartToTakeSnapshot() {
+    // given
+
+    // when
+    logStreamRule.getClock().addTime(Duration.ofMinutes(1));
```
Comment (saig0) [inline logstreams/src/test/java/io/zeebe/logstreams/processor/AsyncSnapshotingTest.java]: Please use the same duration as configured in the director.

File: logstreams/src/test/java/io/zeebe/logstreams/processor/AsyncSnapshotingTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    inOrder.verifyNoMoreInteractions();
+  }
+
+  @Test
+  public void shouldWaitUntilCommitPositionWasUpdated() throws Exception {
+    // given
+    logStreamRule.getClock().addTime(Duration.ofMinutes(1));
+
+    // when
+    waitUntil(() -> asyncSnapshotDirector.lastWrittenEventPosition != INITIAL_POSITION);
```
Comment (saig0) [inline logstreams/src/test/java/io/zeebe/logstreams/processor/AsyncSnapshotingTest.java]: Please avoid using the internal state of the director. This test should also work without this check.

**Review by saig0 — CHANGES_REQUESTED** [review]
I like the idea :+1: Just some comments.

File: logstreams/src/main/java/io/zeebe/logstreams/processor/AsyncSnapshotDirector.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    return name;
+  }
+
+  private String getConditionNameForPosition() {
+    return getName() + "-wait-for-endPosition-committed";
+  }
+
+  @Override
+  protected void onActorCloseRequested() {
+    conditionCheckOut.accept(commitCondition);
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/processor/AsyncSnapshotDirector.java]: This removes the registered condition

[the remark this answers is not in the document]
Comment (saig0) [reply logstreams/src/main/java/io/zeebe/logstreams/processor/AsyncSnapshotDirector.java]: Ok. Do we really need to remove the condition? If should only happens when the logstream is closed.

File: logstreams/src/main/java/io/zeebe/logstreams/processor/StreamProcessor.java [hunk 4]
```diff
@@ -52,6 +54,16 @@ default void onClose() {
...
 
+  /**
+   * Returns the last successful processed event position from the state. This is used after load
+   * the latest snapshot and recover the state, to find the position for reprocessing.
+   *
+   * @return the last successful processed event position from the state
+   */
+  default long getLastSuccessfulProcessedPositionFromState() {
+    return NO_EVENTS_PROCESSED;
+  }
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/processor/StreamProcessor.java]: yes :cry:

File: logstreams/src/main/java/io/zeebe/logstreams/processor/StreamProcessorController.java [hunk 5]
```diff
@@ -174,16 +161,25 @@ protected void onActorStarted() {
...
+    final ZeebeDb zeebeDb = snapshotController.openDb();
+    LOG.info(
+        "Recovered state of partition {} from snapshot at position {}",
+        partitionId,
+        snapshotPosition);
+
+    dbContext = zeebeDb.createContext();
+    streamProcessor = streamProcessorFactory.createProcessor(zeebeDb, dbContext);
+
+    final long snapshotPosition = streamProcessor.getLastSuccessfulProcessedPositionFromState();
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/processor/StreamProcessorController.java]: what do you mean exactly?

File: logstreams/src/main/java/io/zeebe/logstreams/processor/StreamProcessorController.java [hunk 6]
```diff
@@ -296,11 +248,29 @@ protected void onActorClosing() {
...
-            createSnapshot();
-            try {
-              snapshotController.close();
-            } catch (Exception e) {
-              LOG.error("Error on closing snapshotController.", e);
+            final LogStream logStream = streamProcessorContext.logStream;
+
+            if (asyncSnapshotDirector != null) {
+              LOG.info("On closing, will try to enforce snapshot creation.");
+              actor.runOnCompletionBlockingCurrentPhase(
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/processor/StreamProcessorController.java]: So I will remove it ?

File: logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java [hunk 7]
```diff
@@ -110,6 +149,29 @@ public void purgeAll(Predicate<StateSnapshotMetadata> matcher) throws Exception
     }
   }
 
+  @Override
+  public void ensureMaxSnapshotCount(int maxSnapshotCount) throws Exception {
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java]: this means what? :sweat_smile:

[the remark this answers is not in the document]
Comment (saig0) [reply logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java]: #2303

**Review by saig0 — APPROVED** [review]

**Pull request merged**